docs/docs/docs/01-core/utils/traverse-entity.md

```diff
@@ -0,0 +1,142 @@
+
+---
+
+## Overview
+
+### Purpose
+The utility is used for structured data manipulation. It supports various types of attributes, such as relations, components, dynamic zones, and media, allowing seamless traversal and transformation of deeply nested entities.
+
+---
+
+## Function: `traverseEntity`
+
+### Signature
+```typescript
+async function traverseEntity(
+  visitor: [Visitor](#visitoroptions),
+  options: TraverseOptions,
+  entity: Data
+): Promise<Data>;
+```
+
+### Parameters
+- **`visitor`**:  
+  A function executed for each attribute of the entity. It provides context and utilities for modifying the entity.
+
+- **`options`**:  
+  Configuration for traversal, including:
+  - `schema` *(Model)*: The schema definition of the current entity.
+  - `path` *(Path, optional)*: Tracks the current traversal path.
+  - `parent` *(Parent, optional)*: Information about the parent attribute.
+  - `getModel` *(Function)*: A function to retrieve the schema of a model by its UID.
+
+- **`entity`**:  
+  The data entity to be traversed.
+
+### Returns
+A `Promise<Data>` that resolves with the transformed entity.
+
+---
+
+## Related Types
+
+### Visitor
+```typescript
+type Visitor = (visitorOptions: VisitorOptions, visitorUtils: VisitorUtils) => void;
+```
+A function executed during traversal for each attribute.  
+
+**Parameters**:  
+- **`visitorOptions`** *(VisitorOptions)*: Provides details about the current state of traversal.  
+- **`visitorUtils`** *(VisitorUtils)*: Utilities for modifying the entity (e.g., `set`, `remove`).
+
+---
+
+### VisitorOptions
+```typescript
+interface VisitorOptions {
+  data: Record<string, unknown>;
+  schema: Model;
+  key: string;
+  value: Data[keyof Data];
+  attribute?: AnyAttribute;
+  path: Path;
+  getModel(uid: string): Model;
+  parent?: Parent;
+}
+```
+
+- **`data`**: The current entity or its sub-section being processed.  
+- **`schema`**: The schema definition of the current entity.  
+- **`key`**: The current attribute key.  
+- **`value`**: The current attribute value.  
+- **`attribute`** *(optional)*: The schema attribute definition for the current key.  
+- **`path`**: Tracks the traversal path.  
+- **`getModel`**: Function to retrieve a schema definition by UID.  
+- **`parent`** *(optional)*: Information about the parent attribute.
+
+---
+
+### VisitorUtils
+```typescript
+interface VisitorUtils {
+  remove(key: string): void;
+  set(key: string, value: Data): void;
+}
+```
+Utility functions to modify the entity:  
+- **`remove(key)`**: Deletes an attribute from the entity.  
+- **`set(key, value)`**: Sets or updates an attribute on the entity.
+
+---
+
+### Path
+```typescript
+interface Path {
+  raw: string | null;
+  attribute: string | null;
+}
+```
+Tracks the traversal path:
+- **`raw`**: The raw path of the current attribute in the entity.  
+- **`attribute`**: The schema-defined attribute path.
```

Comment:
these 2 are not clear to me, can you add examples?

Reply:
Good catch, I don't completely understand the difference. They both create the key path (eg a.b.c) to the current property being traversed. As far as I can tell the values are always the same. The only way I see they could be different is if the attribute doesn't exist on the schema. In that case the `path.attribute` will not be update whereas `path.raw` will always be updated. Maybe @Convly could clarify the difference?

Reply:
>The only way I see they could be different is if the attribute doesn't exist on the schema. In that case the path.attribute will not be updated whereas path.raw will always be updated

That's basically it. It's made so that you're always sure that path.attribute corresponds to a valid schema attribute whereas path.raw is used to get access to a specific value from the base object. They'll almost always be identical but there are some situation (with nested structures like DZ) where you can have small differences.

tbf, this is mainly used in the other traversal (queries, filters, etc...) when dealing with extra operators or when manually modifying the base object to inject new properties.